File: quicktest/src/cpp/quicktestmain.cpp
// Copyright (C) 2025 The Qt Company Ltd.
// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only

#include "quicktestmain.h"
//#include "rustconv.h"

#include <QtQuickTest>
#include <QQmlEngine>
#include <QQmlContext>

#include <vector>
#include <string>

#include "qtestsetup.h"

namespace rust::bridge
{
    int quickTestMain(const rust::Vec<rust::String>& args, const rust::String &name) {
        return quickTestMainWithSetup(args, name, nullptr);
    }

    int quickTestMainWithProperties(const rust::Vec<rust::String>& args, const rust::String &name, const QVariantMap &properties)
    {
        QtQuickTestSetup setup(properties);
        return quickTestMainWithSetup(args, name, &setup);
    }

    int quickTestMainWithSetup(const rust::Vec<rust::String>& args, const rust::String &name, QObject *setup) {

        std::vector<std::string> storage = {args.begin(), args.end()};
        std::vector<char*> argv;
        for (auto& str : storage)
            argv.push_back(str.data());
        int argc = static_cast<int>(storage.size());

        return quick_test_main_with_setup(argc, argv.data(), std::string(name).data(), nullptr, setup);
    }

} // namespace rust::bridge

File: quicktest/src/quicktestmain.rs
// Copyright (C) 2025 The Qt Company Ltd.
// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only

use qtbridge_type_lib::{QMap_QString_QVariant, QObject};

#[cxx::bridge]
mod ffi {

    unsafe extern "C++" {
        include!("qtbridge-type-lib/src/generated/core/qmap/cpp/qmap_qstring_qvariant.h");
        type QMap_QString_QVariant = super::QMap_QString_QVariant;
        type QObject = super::QObject;
    }


    #[namespace = "rust::bridge"]
    unsafe extern "C++" {
        include!("cpp/quicktestmain.h");

        #[rust_name = "quick_test_main"]
        fn quickTestMain(args: &Vec<String>, name: &String) -> i32;

        #[rust_name = "quick_test_main_with_properties"]
        fn quickTestMainWithProperties(args: &Vec<String>, name: &String, properties: &QMap_QString_QVariant) -> i32;

        #[rust_name = "quick_test_main_with_setup"]
        unsafe fn quickTestMainWithSetup(args: &Vec<String>, name: &String, setup: *mut QObject) -> i32;
    }
}

pub use ffi::quick_test_main;
pub use ffi::quick_test_main_with_setup;
pub use ffi::quick_test_main_with_properties;
